src/verifier.rs

```diff
@@ -0,0 +1,406 @@
+use std::alloc::handle_alloc_error;
+
+use anyhow::Result;
+
+use crate::{
+    AffinePoint, Circuit, Curve, Field, HaloCurve, msm_execute, msm_execute_parallel,
+    msm_precompute, NUM_ROUTED_WIRES, NUM_WIRES, ProjectivePoint, Proof,
+};
+use crate::fft::{fft_precompute, ifft_with_precomputation_power_of_2};
+use crate::partition::get_subgroup_shift;
+use crate::plonk_challenger::Challenger;
+use crate::plonk_gates::evaluate_all_constraints;
+use crate::plonk_proof::OpeningSet;
+use crate::plonk_util::{eval_poly, halo_g, halo_n, pedersen_hash, powers, reduce_with_powers};
+
+const SECURITY_BITS: usize = 128;
+
+pub struct VerificationKey<C: Curve> {
+    selector_commitments: Vec<AffinePoint<C>>,
+    sigma_commitments: Vec<AffinePoint<C>>,
+    degree_log: usize,
+    degree_pow: usize,
+}
+
+pub fn verify_proof_circuit<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+    public_inputs: &[C::ScalarField],
+    proof: &Proof<C>,
+    circuit: &Circuit<C>,
+) -> Result<bool> {
```

Comment:
Maybe we should just use Result<()> so every failure would return an error string? To me it's not obvious what type of error `Ok(false)` would signify.

Reply:
Yes you're right, I'll change it to `Result<()>`.